silkworm/core/protocol/intrinsic_gas.cpp

```diff
@@ -30,6 +30,26 @@ intx::uint128 intrinsic_gas(const UnsignedTransaction& txn, const evmc_revision
         gas += fee::kGTxCreate;
     }
 
+    const uint64_t data_len{txn.data.length()};
+    if (data_len > 0) {
+        const intx::uint128 non_zero_bytes{std::ranges::count_if(txn.data, [](uint8_t c) { return c != 0; })};
+        const intx::uint128 non_zero_gas{rev >= EVMC_ISTANBUL ? fee::kGTxDataNonZeroIstanbul : fee::kGTxDataNonZeroFrontier};
+        gas += non_zero_bytes * non_zero_gas;
+        const intx::uint128 zero_bytes{data_len - non_zero_bytes};
+        gas += zero_bytes * fee::kGTxDataZero;
+
+        // EIP-3860: Limit and meter initcode
+        if (contract_creation && rev >= EVMC_SHANGHAI) {
+            gas += num_words(data_len) * fee::kInitCodeWordCost;
+        }
+
+        if (rev >= EVMC_PRAGUE) {
+            // EIP-7623: Increase calldata cost
+            const intx::uint128 floor_cost = fee::kGTransaction + (zero_bytes + non_zero_bytes * 4) * fee::kTotalCostFloorPerToken;
+            gas = std::max(gas, floor_cost);
```

Comment:
This is not exactly correct. The floor cost is not a part of intrinsic cost. However, the transaction gas limit must be big enough to cover the floor cost just in case.

After the transaction execution you need another modification: if the gas used by the transaction is lower than the floor cost, bump it to the floor cost.